File: web/tests/Player.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { DEFAULT_CONFIG } from '../src/engine/GameConfig';
import { Board } from '../src/engine/Board';
import { PawnState } from '../src/engine/Pawn';
import type { Player } from '../src/engine/Player';

describe('Player', () => {
  let board: Board;
  let player: Player;

  beforeEach(() => {
    board = new Board(DEFAULT_CONFIG);
    player = board.players[0];
  });

  it('renumber all on bench keeps order', () => {
    player.renumber();
    for (let i = 0; i < 4; i++) {
      expect(player.pawns[i].number).toBe(i + 1);
    }
  });

  it('renumber fielder closer to finish gets lower number', () => {
    const corner = board.getCorner(player);
    const far = corner.nextFieldCell!;
    let near = far;
    for (let i = 0; i < 22; i++) {
      near = near.nextFieldCell!;
    }

    const pawnFar = player.pawns[0];
    pawnFar.cell = far;
    pawnFar.state = PawnState.FIELDER;
    far.pawn = pawnFar;

    const pawnNear = player.pawns[1];
    pawnNear.cell = near;
    pawnNear.state = PawnState.FIELDER;
    near.pawn = pawnNear;

    player.renumber();

    expect(pawnNear.number).toBe(1);
    expect(pawnFar.number).toBe(2);
  });

  it('renumber homer beats fielder', () => {
    const corner = board.getCorner(player);
    const fieldCell = corner.nextFieldCell!;
    const homeCell = corner.nextHomeCell!;

    const homer = player.pawns[0];
    homer.cell = homeCell;
    homer.state = PawnState.HOMER;
    homeCell.pawn = homer;

    const fielder = player.pawns[1];
    fielder.cell = fieldCell;
    fielder.state = PawnState.FIELDER;
    fieldCell.pawn = fielder;

    player.renumber();

    expect(homer.number).toBe(1);
    expect(fielder.number).toBe(2);
  });

  it('renumber homer further in home beats homer closer', () => {
    const corner = board.getCorner(player);
    const home1 = corner.nextHomeCell!;
    const home2 = home1.nextHomeCell!;

    const homer1 = player.pawns[0];
    homer1.cell = home1;
    homer1.state = PawnState.HOMER;
    home1.pawn = homer1;

    const homer2 = player.pawns[1];
    homer2.cell = home2;
    homer2.state = PawnState.HOMER;
    home2.pawn = homer2;

    player.renumber();

    expect(homer2.number).toBe(1);
    expect(homer1.number).toBe(2);
  });

  it('renumber newborn after fielder', () => {
    const corner = board.getCorner(player);
    const fieldCell = corner.nextFieldCell!;

    const fielder = player.pawns[0];
    fielder.cell = fieldCell;
    fielder.state = PawnState.FIELDER;
    fieldCell.pawn = fielder;

    const newborn = player.pawns[1];
    newborn.cell = corner;
    newborn.state = PawnState.NEWBORN;
    corner.pawn = newborn;

    player.renumber();

    expect(fielder.number).toBe(1);
    expect(newborn.number).toBe(2);
  });

  it('renumber after move updates numbers', () => {
    const corner = board.getCorner(player);
    const cell1 = corner.nextFieldCell!;
    const cell2 = cell1.nextFieldCell!;

    const pawn1 = player.pawns[0];
    pawn1.cell = cell1;
    pawn1.state = PawnState.FIELDER;
    cell1.pawn = pawn1;

    const pawn2 = player.pawns[1];
    pawn2.cell = cell2;
    pawn2.state = PawnState.FIELDER;
    cell2.pawn = pawn2;

    player.renumber();
    expect(pawn1.number).toBe(2);
    expect(pawn2.number).toBe(1);

    const cell3 = cell2.nextFieldCell!;
    pawn1.moveTo(cell3);
    cell3.pawn = pawn1;
    cell2.pawn = null;

    player.renumber();
    expect(pawn1.number).toBe(1);
    expect(pawn2.number).toBe(2);
  });
});

File: core/src/test/java/ru/anseranser/peshki/engine/PlayerTest.java
package ru.anseranser.peshki.engine;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class PlayerTest {

    private GameConfig config;
    private Board board;
    private Player player;

    @BeforeEach
    void setUp() {
        config = GameConfig.DEFAULT;
        board = new Board(config);
        player = board.getPlayers().get(0);
    }

    @Test
    void renumberAllOnBenchKeepsOrder() {
        player.renumber();
        for (int i = 0; i < 4; i++) {
            assertEquals(i + 1, player.getPawns().get(i).getNumber());
        }
    }

    @Test
    void renumberFielderCloserToFinishGetsLowerNumber() {
        Cell corner = board.getCorner(player);
        Cell far = corner.getNextFieldCell();
        Cell near = far;
        for (int i = 0; i < 22; i++) {
            near = near.getNextFieldCell();
        }

        Pawn pawnFar = player.getPawns().get(0);
        pawnFar.setCell(far);
        pawnFar.setState(Pawn.State.FIELDER);
        far.setPawn(pawnFar);

        Pawn pawnNear = player.getPawns().get(1);
        pawnNear.setCell(near);
        pawnNear.setState(Pawn.State.FIELDER);
        near.setPawn(pawnNear);

        player.renumber();

        assertEquals(1, pawnNear.getNumber());
        assertEquals(2, pawnFar.getNumber());
    }

    @Test
    void renumberHomerBeatsFielder() {
        Cell corner = board.getCorner(player);
        Cell fieldCell = corner.getNextFieldCell();
        Cell homeCell = corner.getNextHomeCell();

        Pawn homer = player.getPawns().get(0);
        homer.setCell(homeCell);
        homer.setState(Pawn.State.HOMER);
        homeCell.setPawn(homer);

        Pawn fielder = player.getPawns().get(1);
        fielder.setCell(fieldCell);
        fielder.setState(Pawn.State.FIELDER);
        fieldCell.setPawn(fielder);

        player.renumber();

        assertEquals(1, homer.getNumber());
        assertEquals(2, fielder.getNumber());
    }

    @Test
    void renumberHomerFurtherInHomeBeatsHomerCloser() {
        Cell corner = board.getCorner(player);
        Cell home1 = corner.getNextHomeCell();
        Cell home2 = home1.getNextHomeCell();

        Pawn homer1 = player.getPawns().get(0);
        homer1.setCell(home1);
        homer1.setState(Pawn.State.HOMER);
        home1.setPawn(homer1);

        Pawn homer2 = player.getPawns().get(1);
        homer2.setCell(home2);
        homer2.setState(Pawn.State.HOMER);
        home2.setPawn(homer2);

        player.renumber();

        assertEquals(1, homer2.getNumber());
        assertEquals(2, homer1.getNumber());
    }

    @Test
    void renumberNewbornAfterFielder() {
        Cell corner = board.getCorner(player);
        Cell fieldCell = corner.getNextFieldCell();

        Pawn fielder = player.getPawns().get(0);
        fielder.setCell(fieldCell);
        fielder.setState(Pawn.State.FIELDER);
        fieldCell.setPawn(fielder);

        Pawn newborn = player.getPawns().get(1);
        newborn.setCell(corner);
        newborn.setState(Pawn.State.NEWBORN);
        corner.setPawn(newborn);

        player.renumber();

        assertEquals(1, fielder.getNumber());
        assertEquals(2, newborn.getNumber());
    }

    @Test
    void renumberFielderOnCornerGetsBestProgress() {
        Cell corner = board.getCorner(player);
        Pawn pawnOnCorner = player.getPawns().get(0);
        pawnOnCorner.setCell(corner);
        pawnOnCorner.setState(Pawn.State.FIELDER);
        corner.setPawn(pawnOnCorner);

        Pawn pawnOnField = player.getPawns().get(1);
        Cell fieldCell = corner.getNextFieldCell();
        pawnOnField.setCell(fieldCell);
        pawnOnField.setState(Pawn.State.FIELDER);
        fieldCell.setPawn(pawnOnField);

        player.renumber();

        assertEquals(1, pawnOnCorner.getNumber());
        assertEquals(2, pawnOnField.getNumber());
    }

    @Test
    void renumberAfterMoveUpdatesNumbers() {
        Cell corner = board.getCorner(player);
        Cell cell1 = corner.getNextFieldCell();
        Cell cell2 = cell1.getNextFieldCell();

        Pawn pawn1 = player.getPawns().get(0);
        pawn1.setCell(cell1);
        pawn1.setState(Pawn.State.FIELDER);
        cell1.setPawn(pawn1);

        Pawn pawn2 = player.getPawns().get(1);
        pawn2.setCell(cell2);
        pawn2.setState(Pawn.State.FIELDER);
        cell2.setPawn(pawn2);

        player.renumber();
        assertEquals(2, pawn1.getNumber());
        assertEquals(1, pawn2.getNumber());

        pawn1.moveTo(cell2.getNextFieldCell());
        cell2.getNextFieldCell().setPawn(pawn1);
        cell2.setPawn(null);

        player.renumber();
        assertEquals(1, pawn1.getNumber());
        assertEquals(2, pawn2.getNumber());
    }
}
